contracts/Cryptoxo.sol

```diff
@@ -0,0 +1,254 @@
+    event showOxoBoardString(
+                            string _lineNum, 
+                            string _firstPosition, 
+                            string _secondPosition, 
+                            string _thirdPosition
+                            );
+                            
+    // modifier for function to return ETH in case 
+    // playerRateProposer made an offer
+    // but none accepted the chellenge
+    modifier onlyPlayerRateProposer() {
+        require(msg.sender == playerRateProposer);
+        _;
+    }
+    
+    // modifier to use forciblyDraw() function 
+    // in case something goes wrong and one of players disappears
+    // to manually draw the game
+    modifier onlyOwner() {
+        require(msg.sender == owner);
+        _;
+    }
+    
+    // according to rules onwer can't play the game
+    // owner can only be a reffery 
+    // in case one of players disappears
+    // to manually draw the game
+    modifier onlyNotOwner() {
+        require(msg.sender != owner);
+        _;
+    }
+    
+    
+    
+    // init
+    function Cryptoxo(){
+        // adding owner in case a contract needs a suicide
+        owner = msg.sender;
+        //movesCounter = 0;
+        
+        // creating an empty oxo board;
+        oxoBoard[0] = ["-", "-", "-"];
+        oxoBoard[1] = ["-", "-", "-"];
+        oxoBoard[2] = ["-", "-", "-"];
+    }
+    
+    
+    
+    
+    function offerRate() public payable onlyNotOwner returns (uint){
+        require(this.balance == msg.value);
+        
+        rateProposed = msg.value;
+        playerRateProposer = msg.sender;
+        
+        return rateProposed;
+    }
+    
+    function replyOffer() public payable onlyNotOwner returns (bool){
+        require(rateProposed != 0);
+        require(msg.sender != playerRateProposer);
+        require(msg.value == rateProposed);
+       
+        players.x = playerRateProposer;
+        players.o = msg.sender;
+        playersVal[playerRateProposer] = "X";
+        playersVal[msg.sender] = "O";
+        playerRateProposer = 0;
+        rateProposed = 0;
+        
+        return true;
+    }
+    
+    function seeOfferRate() public returns (uint){
+        require(rateProposed != 0);
+        return rateProposed;
+    }
+    
+    function makeMove(uint8 _column, uint8 _row) public returns (bool){
+        require(keccak256(playersVal[msg.sender]) != keccak256(0));
+        require(msg.sender != previousMovePlayer);
+        require(keccak256(oxoBoard[_row-1][_column-1]) == keccak256('-'));
+        
+        oxoBoard[_row-1][_column-1] = playersVal[msg.sender];
+        MoveMade(msg.sender, playersVal[msg.sender], _column, _row );
+        
+        if (compareHorizontal() == true) return true;
+        if (compareVertical()) return true;
+        if (compareDiagonal()) return true;
+        
+        movesCounter++;
+        previousMovePlayer = msg.sender;
+        
+        
+        if(movesCounter == 9){
```

Comment:
Is there a situation when someone win at a 9 move?

Reply:
yes, it is possible as far I remember (I'm pretty sure I was able to do it when did the course intro-tests) that last step actually makes someone win instead of draw. I'll try to reproduce it once I come back